web/modules/custom/joinup_core/joinup_core.post_update.php

```diff
@@ -681,3 +681,104 @@ function joinup_core_post_update_install_publication_date() {
   \Drupal::service('module_installer')->install(['joinup_privacy']);
 }
+
+/**
+ * Fix publication dates for all cases.
+ */
+function joinup_core_post_update_fix_all_publication_dateq(&$sandbox) {
+  // This function will iterate through all content and will apply the default
+  // value to all revisions until the first published one and then, it will
+  // apply the created date to all other revisions. When an entity is initially
+  // published, then every subsequent revision should cary the same value
+  // regardless of whether it is published or not.
+  //
+  // In the below example, the entity is created as an unpublished and is
+  // published in the second version. All other versions retain the value.
+  // @codingStandardsIgnoreStart
+  // | revision | status | publication date          |
+  // | 1        | 0      | PUBLICATION_DATE_DEFAULT  |
+  // | 2        | 1      | <actual publication date> |
+  // | 3        | 0      | <actual publication date> |
+  // | 4        | 1      | <actual publication date> |
+  // | 5        | 0      | <actual publication date> |
+  // @codingStandardsIgnoreEnd
+  //
+  // The above approach is applied to the revision table. For the data table,
+  // the published_at value will have 2 cases. It will receive the default value
+  // if the entity has never been published or the publication date if the
+  // entity has been published at least once, regardless of the current state.
+  $database = \Drupal::database();
+  $node_storage = \Drupal::entityTypeManager()->getStorage('node');
+  if (empty($sandbox['nids'])) {
+    // Clean all set publication dates from the revisions table, as the empty
+    // value is needed later on.
+    $database
+      ->update($node_storage->getRevisionDataTable())
+      ->expression('published_at', 'null')
+      ->execute();
+
+    $sandbox['nids'] = $database->query('SELECT nid FROM {node} n')->fetchCol();
+    $sandbox['progress'] = 0;
+    $sandbox['total'] = count($sandbox['nids']);
+  }
+
+  $nids = array_splice($sandbox['nids'], 0, 300);
+  $query = $database->select('node', 'n');
+  $query->leftJoin('node_field_revision', 'nr', 'n.nid = nr.nid');
+  $query->fields('nr', ['nid', 'vid', 'changed']);
+  $query->condition('status', 1);
+  $query->condition('n.nid', $nids, 'IN');
+  $query->groupBy('nr.nid, nr.vid, nr.changed');
+  $query->orderBy('nr.vid', 'DESC');
+
+  $node_data = $query->execute()->fetchAllAssoc('nid');
+  foreach ($nids as $nid) {
+    if (empty($node_data[$nid])) {
+      // Node has never been published. Set the default value to all revisions.
+      $database
+        ->update($node_storage->getDataTable())
+        ->expression('published_at', PUBLICATION_DATE_DEFAULT)
+        ->condition('nid', $nid)
+        ->execute();
+      $database
+        ->update($node_storage->getRevisionDataTable())
+        ->expression('published_at', PUBLICATION_DATE_DEFAULT)
+        ->condition('nid', $nid)
+        ->execute();
+    }
```

Comment:
We can improve performance by keeping track of the node IDs of the nodes that were never published, and then at the end of the batch perform these update queries to all of them at once. This is more efficient than doing two separate queries for each single node.